File: src/ros2_bridge/mappers/foxglove_msgs/text_primitive.rs
//! Typed mapper for `foxglove_msgs/msg/TextPrimitive`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn text_primitive_to_bus(msg: ros_env::foxglove_msgs::msg::TextPrimitive) -> crate::foxglove_msgs::msg::v1::TextPrimitive {
    crate::foxglove_msgs::msg::v1::TextPrimitive {
        pose: Some(crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_bus(msg.pose)),
        billboard: msg.billboard,
        font_size: msg.font_size,
        scale_invariant: msg.scale_invariant,
        color: Some(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus(msg.color)),
        text: crate::ros2_bridge::mappers::convert::from_ros_string(msg.text),
    }
}

pub(crate) fn text_primitive_to_ros(bus: crate::foxglove_msgs::msg::v1::TextPrimitive) -> ros_env::foxglove_msgs::msg::TextPrimitive {
    ros_env::foxglove_msgs::msg::TextPrimitive {
        pose: crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
        billboard: bus.billboard,
        font_size: bus.font_size,
        scale_invariant: bus.scale_invariant,
        color: crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros(bus.color.unwrap_or_default()),
        text: crate::ros2_bridge::mappers::convert::to_ros_string(bus.text),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsTextPrimitiveMapper;

impl TypedTopicMapper for FoxgloveMsgsTextPrimitiveMapper {
    type Ros = ros_env::foxglove_msgs::msg::TextPrimitive;
    type Bus = crate::foxglove_msgs::msg::v1::TextPrimitive;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(text_primitive_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(text_primitive_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/text_primitive.py
"""Generated mapper for `foxglove_msgs/msg/TextPrimitive`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.pose import pose_to_bus, pose_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.color import color_to_bus, color_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import TextPrimitive as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def text_primitive_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    bus.billboard = msg.billboard
    bus.font_size = msg.font_size
    bus.scale_invariant = msg.scale_invariant
    bus.color.CopyFrom(color_to_bus(msg.color))
    bus.text = str(msg.text)
    return bus


def text_primitive_to_ros(bus):
    from foxglove_msgs.msg import TextPrimitive as RosMsg

    out = RosMsg()
    out.pose = pose_to_ros(bus.pose)
    out.billboard = bus.billboard
    out.font_size = bus.font_size
    out.scale_invariant = bus.scale_invariant
    out.color = color_to_ros(bus.color)
    out.text = str(bus.text)
    return out


class FoxgloveMsgsTextPrimitiveMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import TextPrimitive as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return text_primitive_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return text_primitive_to_ros(bus)
